rasa/core/policies/flow_policy.py

```diff
@@ -183,14 +188,147 @@ def _create_prediction_result(
+
+        Args:
+            frame: The frame to push onto the stack.
+        """
+        self.frames.append(frame)
+
+    def update(self, frame: FlowStackFrame) -> None:
+        """Updates the topmost frame.
+
+        Args:
+            frame: The frame to update.
+        """
+        if not self.is_empty():
+            self.pop()
+
+        self.push(frame)
+
+    def advance_top_flow(self, updated_id: Text) -> None:
+        """Updates the topmost flow step.
+
+        Args:
+            updated_id: The updated flow step ID.
+        """
+        if top := self.top():
+            top.step_id = updated_id
+
+    def pop(self) -> FlowStackFrame:
+        """Pops the topmost frame from the stack.
+
+        Returns:
+            The popped frame.
+        """
+        return self.frames.pop()
+
+    def top(self) -> Optional[FlowStackFrame]:
+        """Returns the topmost frame from the stack.
+
+        Returns:
+            The topmost frame.
+        """
+        if self.is_empty():
+            return None
+
+        return self.frames[-1]
+
+    def top_flow(self, flows: FlowsList) -> Optional[Flow]:
+        """Returns the topmost flow from the stack.
+
+        Args:
+            flows: The flows to use.
+
+        Returns:
+            The topmost flow.
+        """
+        if not (top := self.top()):
+            return None
+
+        return flows.flow_by_id(top.flow_id)
+
+    def top_flow_step(self, flows: FlowsList) -> Optional[FlowStep]:
+        """Get the current flow step.
+
+        Returns:
+            The current flow step or `None` if no flow is active."""
+        if not (top := self.top()) or not (top_flow := self.top_flow(flows)):
+            return None
+
+        return top_flow.step_for_id(top.step_id)
+
+    def is_empty(self) -> bool:
+        """Checks if the stack is empty.
+
+        Returns:
+            `True` if the stack is empty, `False` otherwise.
+        """
+        return len(self.frames) == 0
+
+    @staticmethod
+    def from_tracker(tracker: DialogueStateTracker) -> FlowStack:
+        """Creates a `FlowStack` from a tracker.
+
+        Args:
+            tracker: The tracker to create the `FlowStack` from.
+
+        Returns:
+            The created `FlowStack`.
+        """
+        flow_stack = tracker.get_slot(FLOW_STACK_SLOT) or []
+        return FlowStack.from_dict(flow_stack)
+
+
+@dataclass
+class ActionPrediction:
+    """Represents an action prediction."""
+
+    action_name: Optional[Text]
+    """The name of the predicted action."""
+    score: float
+    """The score of the predicted action."""
+
+
+@dataclass
+class FlowStackFrame:
     """Represents the current flow step."""
 
     flow_id: Text
-    step_id: Optional[Text] = None
+    """The ID of the current flow."""
+    step_id: Text = START_STEP
+    """The ID of the current step."""
+    return_to_loop: Optional[Text] = None
+    """The name of the loop that should be actived when this stack frame ends."""
```

Comment:
i did remove this now, but included an enum to label the stack with a type. some storage needs to be added, otherwise we don't know that we got interrupted rather than linked.